[pytree] freeze attributes of `TreeSpec`

XuehaiPan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114946

This PR adds `frozen=True` to the `@dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

XuehaiPan committed:
[pytree] freeze attributes of `TreeSpec`

[ghstack-poisoned]

Skylion007 commented:
We should also consider adding `slots=True` for performance reasons if possible.

XuehaiPan commented:
> We should also consider adding `slots=True` for performance reasons if possible.

I tried this, but it seems there are some upstream bugs for the `slots` option in `@dataclass`.

- python/cpython#105936
- python/cpython#111500

```python
In [1]: import dataclasses

In [2]: @dataclasses.dataclass(frozen=True)
   ...: class A:
   ...:     a: int
   ...:     

In [3]: a = A(1)

In [4]: a.x = 1
FrozenInstanceError: cannot assign to field 'x'

In [5]: @dataclasses.dataclass(slots=True)
   ...: class B:
   ...:     b: int
   ...:     

In [6]: b = B(1)

In [7]: b.x = 1
AttributeError: 'B' object has no attribute 'x'

In [8]: @dataclasses.dataclass(frozen=True, slots=True)
   ...: class C:
   ...:     c: int
   ...:     

In [9]: c = C(1)

In [10]: c.x = 1
TypeError: super(type, obj): obj must be an instance or subtype of type
```

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519

[ghstack-poisoned]

XuehaiPan commented on torch/utils/_pytree.py:
```diff
@@ -470,14 +470,23 @@ def _is_leaf(tree: PyTree) -> bool:
+        num_nodes = 1 + sum(spec.num_nodes for spec in self.children_specs)
+        num_leaves = sum(spec.num_leaves for spec in self.children_specs)
+        num_children = len(self.children_specs)
+        object.__setattr__(self, "num_nodes", num_nodes)
+        object.__setattr__(self, "num_leaves", num_leaves)
+        object.__setattr__(self, "num_children", num_children)
```
Dynamo does not support `object.__setattr__(obj, name, value)` yet.

For all frozen dataclasses, the assignment statement in `__init__` are `object.__setattr__(obj, name, value)` rather than `obj.<name> = value`.

https://github.com/python/cpython/blob/05a370abd6cdfe4b54be60b3b911f3a441026bb2/Lib/dataclasses.py#L477-L486

```python
def _field_assign(frozen, name, value, self_name):
    # If we're a frozen class, then assign to our fields in __init__
    # via object.__setattr__.  Otherwise, just use a simple
    # assignment.
    #
    # self_name is what "self" is called in this function: don't
    # hard-code "self", since that might be a field name.
    if frozen:
        return f'__dataclass_builtins_object__.__setattr__({self_name},{name!r},{value})'
    return f'{self_name}.{name}={value}'
```

The error:

```text
torch._dynamo.exc.Unsupported: call_method BuiltinVariable(object) __setattr__ [UserDefinedObjectVariable(TreeSpec), ConstantVariable(str), ConstantVariable(int)] {}
```

A minimal script to reproduce:

```python
import dataclasses

import torch


@dataclasses.dataclass
class A:
    a: int


def fn() -> None:
    a = A(1)
    object.__setattr__(a, "a", 2)


comp_out = torch._dynamo.optimize(nopython=True)(fn)()
```

XuehaiPan commented on torch/utils/_pytree.py:
```diff
@@ -470,14 +470,23 @@ def _is_leaf(tree: PyTree) -> bool:
 # children_specs: specs for each child of the root Node
 # num_leaves: the number of leaves
-@dataclasses.dataclass
+@dataclasses.dataclass(init=True, frozen=True, eq=True, repr=False)
```
Dynamo does not support frozen dataclasses.

It errors for statement:

```python
return TreeSpec(node_type, context, children_specs)
```

The error:

```text
torch._dynamo.exc.Unsupported: call_function UserDefinedClassVariable() [BuiltinVariable(list), ConstantVariable(NoneType), ListVariable()] {}
```

A minimal script to reproduce:

```python
import dataclasses

import torch


@dataclasses.dataclass(frozen=True)
class A:
    a: int


def fn() -> None:
    a = A(1)


comp_out = torch._dynamo.optimize(nopython=True)(fn)()
```

The error for statement `a = A(1)`:

```text
torch._dynamo.exc.Unsupported: call_function UserDefinedClassVariable() [ConstantVariable(int)] {}
```

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519 voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519 voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519 voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519 voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519 voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519 voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] freeze attributes of `TreeSpec`"


This PR adds `frozen=True` to the `dataclasses.dataclass` decorator. This makes all attributes for `TreeSpec` object read-only.


cc zou3519 voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

XuehaiPan commented:
@pytorchbot rebase
